framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C

```diff
@@ -0,0 +1,213 @@
+          continue;
+
+        points[i_from].push_back(*node + _from_positions[i_from]);
+        auto dof = node->dof_number(from_sys.number(), from_var_num, 0);
+
+        local_values[i_from].push_back((*from_sys.solution)(dof));
+      }
+    }
+    else
+    {
+      if (!from_boundaries.empty())
+      {
+        mooseError("You can not restrict an elemental variable to boundary");
+      }
+      for (auto & elem : as_range(from_mesh.getMesh().local_elements_begin(),
+                                  from_mesh.getMesh().local_elements_end()))
+      {
+        if (elem->n_dofs(from_sys.number(), from_var_num) < 1)
+          continue;
+
+        if (!from_blocks.empty() && !hasBlocks(from_blocks, elem))
+          continue;
+
+        auto dof = elem->dof_number(from_sys.number(), from_var_num, 0);
+        points[i_from].push_back(elem->vertex_average() + _from_positions[i_from]);
+        local_values[i_from].push_back((*from_sys.solution)(dof));
+      }
+    }
+
+    std::shared_ptr<KDTree> _kd_tree =
+        std::make_shared<KDTree>(points[i_from], from_mesh.getMaxLeafSize());
+    kdtrees[i_from] = _kd_tree;
+  }
+}
+
+void
+MultiAppGeneralFieldNearestNodeTransfer::evaluateInterpValues(
+    const std::vector<Point> & incoming_points, std::vector<std::pair<Real, Real>> & outgoing_vals)
+{
+  evaluateInterpValuesNearestNode(
+      _local_kdtrees, _local_points, _local_values, incoming_points, outgoing_vals);
+}
+
+void
+MultiAppGeneralFieldNearestNodeTransfer::evaluateInterpValuesNearestNode(
+    const std::vector<std::shared_ptr<KDTree>> & local_kdtrees,
+    const std::vector<std::vector<Point>> & local_points,
+    const std::vector<std::vector<Real>> & local_values,
+    const std::vector<Point> & incoming_points,
+    std::vector<std::pair<Real, Real>> & outgoing_vals)
+{
+  dof_id_type i_pt = 0;
+  for (auto & pt : incoming_points)
+  {
+    // Loop until we've found the lowest-ranked app that actually contains
+    // the quadrature point.
+    for (MooseIndex(_from_problems.size()) i_from = 0;
+         i_from < _from_problems.size() &&
+         outgoing_vals[i_pt].first == GeneralFieldTransfer::BetterOutOfMeshValue;
+         ++i_from)
+    {
+
+      std::vector<std::size_t> return_index(_num_nearest_points);
+      std::vector<Real> return_dist_sqr(_num_nearest_points);
+      if (local_kdtrees[i_from]->numberCandidatePoints())
+      {
+        local_kdtrees[i_from]->neighborSearch(
+            pt, _num_nearest_points, return_index, return_dist_sqr);
+        Real val_sum = 0, dist_sum = 0;
+        for (auto index : return_index)
+        {
+          val_sum += local_values[i_from][index];
+          dist_sum += (local_points[i_from][index] - pt).norm();
+        }
+        // Use mesh function to compute interpolation values
+        // Assign value
+        outgoing_vals[i_pt] = {val_sum / return_index.size(), dist_sum / return_dist_sqr.size()};
```

Comment:
This seems like the most likely source of trouble on distributed meshes.  Every query is returning the average value from N nearest source points, as prioritized by the average distance from N nearest source points ... what are we doing to ensure that we aren't prioritizing query responses with missing points?  If processor A detects that the closest 4 points are (0.5,0.6,0.7,0.8) units away, but processor B only sees the 0.5-away point, aren't we going to then prefer processor B's value calculation?

Reply:
Resolution on this is that you are right but there's no way around this, save for communicating multiple points from all the origins for each target